adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,261 @@
+* The `runtime` module should focus on the internal implementation and runtime logic, delegating public APIs to the `api` module.
+* The `runtime-api` module contains both the public API and the SPI, allowing other extensions to depend on it without pulling in the full extension.
+
+For extensions requiring tight-coupling, the `runtime` module can still be used to expose public APIs, but this should be avoided when possible.
+
+[cols="1,3,2",options="header"]
+|===
+| Scenario | Recommended Structure | Notes
+
+| Small/simple extension not reused by others
+| `runtime`
+| Keep everything in `runtime`. Avoid unnecessary modularity.
+
+| Extension exposes public types used by application code or other extensions
+| `runtime-api` + `runtime`
+| Split APIs (annotations, interfaces, utility classes) into `runtime-api`. Keep internal logic in `runtime`.
+
+| Extension contributes dev-mode-only logic (e.g., Dev UI)
+| `runtime` + `runtime-dev`
+| Add `runtime-dev` to isolate dev-only classes. Avoid shipping to production.
+
+| Extension defines types meant to be implemented by others (e.g., customizers, listeners)
+| `runtime-api` + `runtime`
+| Consider these types part of the SPI. Place them in `runtime-api` to allow loose coupling.
+
+| Extension wants to enforce tight coupling (full extension must be present)
+| `runtime` only
+| Expose public types directly from `runtime`. Use cautiously; limits flexibility and reuse.
+
+| Extension depends on or provides code generation
+| `runtime` + `codegen`
+| Keep codegen logic isolated. Optional, depending on feature set.
+|===
+
+=== Package Name Rules
+
+Extensions must use a well-defined package structure to avoid split packages.
+
+[NOTE]
+====
+In this section, rules are given for the `io.quarkus` and `io.quarkiverse` namespaces.
+When the root is different, the rules apply similarly, replacing `io.quarkus` or `io.quarkiverse` with the appropriate root package.
+For example: `org.apache.camel.quarkus.component` would use `org.apache.camel.quarkus.component.runtime`, `org.apache.camel.quarkus.component.runtime.dev`, etc.
+====
+
+==== runtime module
+* `io.<quarkus|quarkiverse>.<extension-name>.runtime`: Internal implementation. Not part of the public API. May include subpackages (excluding dev).
+* `io.<quarkus|quarkiverse>.<extension-name>.runtime.graal`: GraalVM substitutions. Not part of the public API.
+* `io.<quarkus|quarkiverse>.<extension-name>`: Public API when requiring tight-coupling. May include subpackages (excluding `api`). Example: `io.quarkus.cache`. Note that this is discouraged in favor of the `runtime-api` module.
+
+
+==== deployment module
+* `io.<quarkus|quarkiverse>.<extension-name>.deployment`: Internal build logic (processors, build steps, internal build items). Public APIs must reside in the `deployment-api` module. May include subpackages (excluding `api`).
```

Comment:
I wonder if for the deployment too, we should be clear about what's internal.

It would make it clear which build items are supposed to be consumed by others.